src/transformers/generation_utils.py

```diff
@@ -1645,6 +1749,237 @@ def typeerror():
+                if this_peer_finished_flag.item() == 0.0:
+                    break
+
+            # contrastive search decoding consists of two steps: (1) candidate tokens recall; (2) candidate re-rank by degeneration penalty
+            (
+                next_tokens,
+                past_key_values,
+                last_hidden_states,
+                logit_for_next_step,
+                selected_scores,
+                decoder_hidden_states_one_step,
+            ) = ContrastiveDecodingOneStepFast(
+                self,
+                beam_width=top_k,
+                penalty_alpha=penalty_alpha,
+                last_hidden_states=last_hidden_states,
+                logit_for_next_step=logit_for_next_step,
+                is_encoder_decoder=self.config.is_encoder_decoder,
+                **model_inputs,
+            )
+
+            if synced_gpus and this_peer_finished:
+                continue  # don't waste resources running the code we don't need
+
+            # Store scores, attentions and hidden_states when required
+            if return_dict_in_generate:
+                if output_scores:
+                    scores += (selected_scores,)
+
+                if output_hidden_states:
+                    decoder_hidden_states += (decoder_hidden_states_one_step,)
+
+            # finished sentences should have their next token be a padding token
+            if eos_token_id is not None:
+                if pad_token_id is None:
+                    raise ValueError("If `eos_token_id` is defined, make sure that `pad_token_id` is defined.")
+                next_tokens = next_tokens * unfinished_sequences + pad_token_id * (1 - unfinished_sequences)
+
+            # update generated ids, model inputs, and length for next step
+            input_ids = torch.cat([input_ids, next_tokens[:, None]], dim=-1)
+            if self.config.is_encoder_decoder:
+                outputs = Seq2SeqLMOutput(
+                    past_key_values=past_key_values,
+                )
+            else:
+                outputs = CausalLMOutputWithCrossAttentions(
+                    past_key_values=past_key_values, attentions=model_kwargs["attention_mask"]
+                )
+            model_kwargs = self._update_model_kwargs_for_generation(
+                outputs, model_kwargs, is_encoder_decoder=self.config.is_encoder_decoder
+            )
+
+            # if eos_token was found in one sentence, set sentence to finished
+            if eos_token_id is not None:
+                unfinished_sequences = unfinished_sequences.mul((next_tokens != eos_token_id).long())
+
+            # stop when each sentence is finished, or if we exceed the maximum length
+            if unfinished_sequences.max() == 0 or stopping_criteria(input_ids, scores):
+                if not synced_gpus:
+                    break
+                else:
+                    this_peer_finished = True
+
+            # prepare model inputs
```

Comment:
Not a huge fan of having this down here as it's different from the logic of the other generate methods. Could we somehow move it up? Potentially skipping the very first step as it's already done? At the moment this will always be run once unnecessarily at the end